File: scripts/validate-specs.mjs
#!/usr/bin/env node
/**
 * Documentation validator (docs/spec/10-testing.md).
 *
 * Fails the build when:
 *  1. a requirement ID is defined twice,
 *  2. an `auto` requirement has no referencing test,
 *  3. a relative markdown link in docs/ or README/CLAUDE files is broken,
 *  4. openapi.yaml doesn't parse, or implemented Fastify routes and documented
 *     paths disagree (API-003).
 */
import { readdir, readFile } from 'node:fs/promises'
import { existsSync } from 'node:fs'
import path from 'node:path'

const root = path.resolve(import.meta.dirname, '..')
const errors = []

async function walk(dir, filter) {
  if (!existsSync(dir)) return []
  const out = []
  for (const entry of await readdir(dir, { withFileTypes: true })) {
    if (entry.name === 'node_modules' || entry.name === '.git' || entry.name === 'dist') continue
    const full = path.join(dir, entry.name)
    if (entry.isDirectory()) out.push(...(await walk(full, filter)))
    else if (filter(full)) out.push(full)
  }
  return out
}

// ---- 1 & 2: requirement tables vs test references ----------------------------------
const REQ_ROW = /^\|\s*([A-Z]+-\d{3})\s*\|(.+)\|\s*(auto|manual)\s*\|\s*$/
const specFiles = await walk(path.join(root, 'docs'), (f) => f.endsWith('.md'))
const requirements = new Map() // id -> {file, verify}
for (const file of specFiles) {
  for (const line of (await readFile(file, 'utf8')).split('\n')) {
    const m = line.match(REQ_ROW)
    if (!m) continue
    const [, id, , verify] = m
    if (requirements.has(id)) {
      errors.push(`Duplicate requirement ID ${id} (${path.relative(root, file)} and ${requirements.get(id).file})`)
    } else {
      requirements.set(id, { file: path.relative(root, file), verify })
    }
  }
}
if (requirements.size === 0) errors.push('No requirement tables found under docs/ — spec drift?')

const testFiles = [
  ...(await walk(path.join(root, 'test'), (f) => f.endsWith('.ts'))),
  ...(await walk(path.join(root, 'scripts'), (f) => f.endsWith('.ts') || f.endsWith('.mjs'))),
]
const testCorpus = (await Promise.all(testFiles.map((f) => readFile(f, 'utf8')))).join('\n')

for (const [id, meta] of requirements) {
  if (meta.verify !== 'auto') continue
  if (!testCorpus.includes(id)) {
    errors.push(`Requirement ${id} (${meta.file}) is marked auto but no test references it`)
  }
}

// ---- 3: relative markdown links -----------------------------------------------------
const mdFiles = [
  ...specFiles,
  ...['README.md', 'CLAUDE.md'].map((f) => path.join(root, f)).filter((f) => existsSync(f)),
]
const LINK = /\[[^\]]*\]\(([^)]+)\)/g
for (const file of mdFiles) {
  const text = await readFile(file, 'utf8')
  for (const m of text.matchAll(LINK)) {
    const target = m[1]
    if (/^(https?:|mailto:|#)/.test(target)) continue
    const resolved = path.resolve(path.dirname(file), target.split('#')[0])
    if (!existsSync(resolved)) errors.push(`Broken link in ${path.relative(root, file)}: ${target}`)
  }
}

// ---- 4: OpenAPI ↔ implemented routes ------------------------------------------------
try {
  const { parse } = await import('yaml')
  const doc = parse(await readFile(path.join(root, 'docs/spec/openapi.yaml'), 'utf8'))
  const documented = new Set()
  for (const [p, methods] of Object.entries(doc.paths ?? {})) {
    for (const method of Object.keys(methods)) {
      if (['get', 'post', 'put', 'patch', 'delete'].includes(method)) {
        documented.add(`${method.toUpperCase()} ${p.replace(/\{[^}]+\}/g, ':param')}`)
      }
    }
  }

  // Build the app with a route-collection hook; the pool is never connected.
  const { createPool } = await import('../src/db.js')
  const { buildApp } = await import('../src/app.js')
  const implemented = new Set()
  const pool = createPool('postgres://unused:unused@127.0.0.1:1/unused')
  const app = await buildApp({
    pool,
    onRoute: (method, url) => implemented.add(`${method} ${url.replace(/:[^/]+/g, ':param')}`),
  })
  await app.close()

  for (const r of implemented) {
    if (!documented.has(r)) errors.push(`Route implemented but not in openapi.yaml: ${r}`)
  }
  for (const r of documented) {
    if (!implemented.has(r)) errors.push(`Route documented in openapi.yaml but not implemented: ${r}`)
  }
} catch (err) {
  errors.push(`OpenAPI validation failed to run: ${err.message}`)
}

// ---- report -------------------------------------------------------------------------
const autoCount = [...requirements.values()].filter((r) => r.verify === 'auto').length
if (errors.length > 0) {
  console.error(`Spec validation FAILED with ${errors.length} problem(s):\n`)
  for (const e of errors) console.error(`  ✗ ${e}`)
  process.exit(1)
}
console.log(
  `Spec validation OK: ${requirements.size} requirements (${autoCount} auto, all test-covered), ` +
    `${mdFiles.length} docs checked, OpenAPI and routes in sync.`,
)

File: .claude/skills/validate-specs/validate_specs.py
#!/usr/bin/env python3
"""Portable spec/requirements validator — stdlib only, runs anywhere.

Spec-driven repos declare behavior as requirement rows in their spec docs:

    | AREA-NNN | short description | auto |

The third column is the **verification tag**, one of `auto | manual | planned`
(this is the canonical vocabulary — not "Testable"/"Manual"). This validator
enforces three core gates that keep the spec honest across a JS, Kotlin,
Python, C#, or docs-only repo without any language-specific tooling:

  1. **Unique IDs** — a requirement ID defined twice is an error.
  2. **auto ⇒ covered** — every `auto` requirement must be referenced *by name*
     in at least one test file under the configured test dir(s). The reference
     must be meaningful: the exact ID, on a word boundary, inside a file that
     actually lives under a test directory. This fixes the weakness in the
     original roadtrip validators, where a bare substring anywhere in the
     concatenated corpus counted (so `LOC-001` was "covered" by an unrelated
     `LOC-0012`, or by a stray mention in prose).
  3. **Links resolve** — every relative markdown link in the spec docs points
     at a file that exists on disk.

`manual` requirements are verified by a human and need no test reference;
`planned` requirements are not built yet and are reported but not enforced.

An optional fourth gate (API-contract diff, e.g. OpenAPI vs. implemented
routes) is language-specific and stays in each repo's own richer validator;
this portable core is the shared floor every repo can run.

Usage:
    validate_specs.py [--spec-dir docs/spec] \
        [--test-dir DIR ...] [--suffix .ts ...] [--quiet]

Exit code 0 on success, 1 on any gate failure.
"""
from __future__ import annotations

import argparse
import os
import re
import sys

# Requirement row: | AREA-001 | ... | auto |   (AREA is 1+ upper-case letters
# and/or digits after the first letter; NNN is 3+ digits).
REQ_ROW = re.compile(
    r"^\|\s*([A-Z][A-Z0-9]*-\d{3,})\s*\|.*\|\s*(auto|manual|planned)\s*\|\s*$"
)
LINK = re.compile(r"\[[^\]]*\]\(([^)]+)\)")
DEFAULT_SUFFIXES = (
    ".ts", ".tsx", ".mjs", ".js", ".kt", ".kts", ".py", ".java", ".cs", ".go",
)
SKIP_DIRS = {".git", "node_modules", "dist", "build", ".gradle", "__pycache__"}


def _walk(root, predicate):
    out = []
    if not os.path.isdir(root):
        return out
    for dirpath, dirs, files in os.walk(root):
        dirs[:] = [d for d in dirs if d not in SKIP_DIRS]
        for name in sorted(files):
            full = os.path.join(dirpath, name)
            if predicate(full):
                out.append(full)
    return out


def resolve_spec_dir(spec_dir):
    """Use spec_dir if it exists; else fall back to docs/ then '.'."""
    for candidate in (spec_dir, "docs", "."):
        if candidate and os.path.isdir(candidate):
            return candidate
    return spec_dir


def collect_requirements(spec_dir):
    """Return (requirements, errors). requirements: id -> (file, verify)."""
    errors = []
    requirements = {}
    md_files = _walk(spec_dir, lambda f: f.endswith(".md"))
    for file in md_files:
        with open(file, "r", encoding="utf-8") as fh:
            for line in fh:
                m = REQ_ROW.match(line.rstrip("\n"))
                if not m:
                    continue
                rid, verify = m.group(1), m.group(2)
                if rid in requirements:
                    errors.append(
                        f"Duplicate requirement ID {rid} "
                        f"({file} and {requirements[rid][0]})"
                    )
                else:
                    requirements[rid] = (file, verify)
    if not requirements:
        errors.append(f"No requirement rows found under {spec_dir} — spec drift?")
    return requirements, errors


def build_test_index(test_dirs, suffixes):
    """Map each test file under the configured dirs to its text."""
    index = {}
    for d in test_dirs:
        for f in _walk(d, lambda p: p.endswith(tuple(suffixes))):
            try:
                with open(f, "r", encoding="utf-8", errors="replace") as fh:
                    index[f] = fh.read()
            except OSError:
                continue
    return index


def _meaningful_ref(rid, text):
    # Exact ID on a word boundary: LOC-001 must not match LOC-0012 or XLOC-001.
    pattern = r"(?<![A-Za-z0-9])" + re.escape(rid) + r"(?![A-Za-z0-9])"
    return re.search(pattern, text) is not None


def check_coverage(requirements, test_index):
    errors = []
    for rid, (file, verify) in sorted(requirements.items()):
        if verify != "auto":
            continue
        if not any(_meaningful_ref(rid, text) for text in test_index.values()):
            errors.append(
                f"Requirement {rid} ({file}) is marked auto but no test file "
                f"under the configured test dir(s) references it"
            )
    return errors


def check_links(spec_dir):
    errors = []
    for file in _walk(spec_dir, lambda f: f.endswith(".md")):
        with open(file, "r", encoding="utf-8") as fh:
            text = fh.read()
        for m in LINK.finditer(text):
            target = m.group(1).strip()
            if re.match(r"^(https?:|mailto:|tel:|#)", target):
                continue
            resolved = os.path.normpath(
                os.path.join(os.path.dirname(file), target.split("#")[0])
            )
            if not os.path.exists(resolved):
                errors.append(f"Broken link in {file}: {target}")
    return errors


def run(spec_dir, test_dirs, suffixes):
    spec_dir = resolve_spec_dir(spec_dir)
    errors = []
    requirements, req_errors = collect_requirements(spec_dir)
    errors += req_errors

    test_index = build_test_index(test_dirs, suffixes)
    if not test_index and any(v == "auto" for _, v in requirements.values()):
        errors.append(
            "No test files found under the configured test dir(s) "
            f"({', '.join(test_dirs) or '<none>'}) — cannot verify auto coverage"
        )
    errors += check_coverage(requirements, test_index)
    errors += check_links(spec_dir)
    return requirements, errors


def main(argv=None):
    ap = argparse.ArgumentParser(description=__doc__,
                                 formatter_class=argparse.RawDescriptionHelpFormatter)
    ap.add_argument("--spec-dir", default="docs/spec",
                    help="Directory holding requirement tables (default docs/spec, "
                         "falling back to docs/ then .).")
    ap.add_argument("--test-dir", action="append", default=[], dest="test_dirs",
                    help="Directory of test sources to scan for ID references "
                         "(repeatable).")
    ap.add_argument("--suffix", action="append", default=[], dest="suffixes",
                    help="Test file suffix to include (repeatable; default covers "
                         "common test languages).")
    ap.add_argument("--quiet", action="store_true")
    args = ap.parse_args(argv)

    suffixes = tuple(args.suffixes) if args.suffixes else DEFAULT_SUFFIXES
    requirements, errors = run(args.spec_dir, args.test_dirs, suffixes)

    if errors:
        sys.stderr.write(f"Spec validation FAILED with {len(errors)} problem(s):\n")
        for e in errors:
            sys.stderr.write(f"  x {e}\n")
        return 1

    if not args.quiet:
        auto = sum(1 for _, v in requirements.values() if v == "auto")
        manual = sum(1 for _, v in requirements.values() if v == "manual")
        planned = sum(1 for _, v in requirements.values() if v == "planned")
        print(
            f"Spec validation OK: {len(requirements)} requirements "
            f"({auto} auto all test-covered, {manual} manual, {planned} planned); "
            f"links resolve."
        )
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
